pkgs/build-support/dotnet/build-dotnet-module/hooks/dotnet-fixup-hook.sh

```diff
@@ -30,7 +30,7 @@ wrapDotnetProgram() {
 dotnetFixupHook() {
     echo "Executing dotnetFixupPhase"
 
-    if [ "${executables-}" ]; then
+    if [[ "$(declare -p executables)" =~ "declare -a" ]]; then
```

Comment:
Oh right, my bad, I misunderstood what this line did.

Reply:
Could we just check `[[ -v executables ]]`?  That should be false with `null`.

Reply:
> Oh right, my bad, I misunderstood what this line did.

I can add a comment to make it clearer, it's non-intuitive.

> Could we just check `[[ -v executables ]]`? That should be false with `null`.

A lot of things are false with `null`, unfortunately most of them are also false with an empty array. I spent a fair amount of time trying to find a test that returns true for filled arrays and empty arrays, but false on null, and I landed on this.

Reply:
I just tested it like this:

```
$ nix-shell --expr 'with import ./. {}; stdenv.mkDerivation { name="foo"; a = null; b = []; c = [ "foo" ]; __structuredAttrs = true; }' --run 'declare -p a b c'
/tmp/nix-shell-67841-0/rc: line 30: declare: a: not found
declare -a b=()
declare -a c=([0]="foo")
$ nix-shell --expr 'with import ./. {}; stdenv.mkDerivation { name="foo"; a = null; b = []; c = [ "foo" ]; }' --run 'declare -p a b c'
/tmp/nix-shell-68373-0/rc: line 3: declare: a: not found
declare -x b=""
declare -x c="foo"
```

So in both cases, `[[ -v executables ]]` should be false for null, and true for both empty and non-empty lists.

Reply:
Oh, I thought you were suggesting to use `[[ -v` but still keep `__structuredAttrs = true;`. This only works with `__structuredAttrs = false`, but maybe that's acceptable?

```
$ nix-shell --expr 'with import ./. {}; stdenv.mkDerivation { name="foo"; a = null; b = []; c = [ "foo" ]; }' --run 'declare -p a b c; [[ -v a ]] && echo "a non-null" || echo "a null"; [[ -v b ]] && echo "b non-null" || echo "b null"; [[ -v c ]] && echo "c non-null" || echo "c null"'
/tmp/nix-shell-1418109-0/rc: line 3: declare: a: not found
declare -x b=""
declare -x c="foo"
a null
b non-null
c non-null

$ nix-shell --expr 'with import ./. {}; stdenv.mkDerivation { name="foo"; a = null; b = []; c = [ "foo" ]; __structuredAttrs = true; }' --run 'declare -p a b c; [[ -v a ]] && echo "a non-null" || echo "a null"; [[ -v b ]] && echo "b non-null" || echo "b null"; [[ -v c ]] && echo "c non-null" || echo "c null"'
structuredAttrs is enabled
/tmp/nix-shell-1427918-0/rc: line 30: declare: a: not found
declare -a b=()
declare -a c=([0]="foo")
a null
b null
c non-null
```

Reply:
Well at least this means I don't have to set `__structuredAttrs` by default, but on the other hand it feels like this is installing a footgun :/

Reply:
I mean, we currently have very strict requirements on which values go on `executables`, so this should be safe, at least *for now*.

And it's a better option than breaking some builds (`__structuredAttrs`) for something I'm not sure is a common use case (empty executables).

Reply:
> This only works with __structuredAttrs = false,

Ah, sorry, I thought I tested that case.

In that case it might be best to do `if declare -p executables &>/dev/null; then` ?  That one does seem to work for both strings and arrays.

I'm not fundamentally against `__structuredAttrs`; it just seems risky.

Reply:
```
$ nix-shell --expr 'with import ./. {}; stdenv.mkDerivation { name="foo"; a = null; b = []; c = [ "foo" ]; }' --run 'declare -p a b c; if declare -p a &>/dev/null; then echo "a non-null"; else echo "a null"; fi; if declare -p b &>/dev/null; then echo "b non-null"; else echo "b null"; fi; if declare -p c &>/dev/null; then echo "c non-null"; else echo "c null"; fi'
/tmp/nix-shell-2079752-0/rc: line 3: declare: a: not found
declare -x b=""
declare -x c="foo"
a null
b non-null
c non-null

$ nix-shell --expr 'with import ./. {}; stdenv.mkDerivation { name="foo"; a = null; b = []; c = [ "foo" ]; __structuredAttrs = true; }' --run 'declare -p a b c; if declare -p a &>/dev/null; then echo "a non-null"; else echo "a null"; fi; if declare -p b &>/dev/null; then echo "b non-null"; else echo "b null"; fi; if declare -p c &>/dev/null; then echo "c non-null"; else echo "c null"; fi'
structuredAttrs is enabled
/tmp/nix-shell-2080895-0/rc: line 30: declare: a: not found
declare -a b=()
declare -a c=([0]="foo")
a null
b non-null
c non-null
```

Great, seems to work!

Reply:
The issue is that `__structuredAttrs` is what broke the other 3 tests, many things need to be adapted for us to switch to it.

I think the reasonable path to go here is use `[ -v` and then we start work on switching to `__structuredAttr` later on, otherwise this PR will become huge because of unrelated changes to the original issue.